llvm/lib/Target/AMDGPU/SIISelLowering.cpp

```diff
@@ -1831,26 +1831,17 @@ bool SITargetLowering::allowsMisalignedMemoryAccessesImpl(
            Subtarget->hasUnalignedDSAccessEnabled();
   }
 
-  if (AddrSpace == AMDGPUAS::PRIVATE_ADDRESS) {
-    bool AlignedBy4 = Alignment >= Align(4);
-    if (IsFast)
-      *IsFast = AlignedBy4;
-
-    return AlignedBy4 ||
-           Subtarget->enableFlatScratch() ||
-           Subtarget->hasUnalignedScratchAccess();
-  }
-
   // FIXME: We have to be conservative here and assume that flat operations
   // will access scratch.  If we had access to the IR function, then we
   // could determine if any private memory was used in the function.
-  if (AddrSpace == AMDGPUAS::FLAT_ADDRESS &&
-      !Subtarget->hasUnalignedScratchAccess()) {
+  if (AddrSpace == AMDGPUAS::PRIVATE_ADDRESS ||
+      AddrSpace == AMDGPUAS::FLAT_ADDRESS) {
     bool AlignedBy4 = Alignment >= Align(4);
     if (IsFast)
       *IsFast = AlignedBy4;
 
-    return AlignedBy4;
+    return AlignedBy4 || Subtarget->enableFlatScratch() ||
```

Comment:
enableFlatScratch should be checked, the scratch instructions still should respect whether the unaligned access mode is enabled

Reply:
Note that we don't do that in the current trunk state: see the old [line 1840](https://github.com/llvm/llvm-project/pull/110219/files#diff-e78d2fbd64648d787707fd3d4e7e5b5d2f00fb9c09972937718a11237933c597L1840). Removing `Subtarget->enableFlatScratch()` from the OR would make code generation more pessimistic for cases where unaligned scratch is not explicitly enabled but flat scratch is, e.g., gfx12. Should we change that anyway?

Reply:
I think this is just a bug. The question of whether we should assume unaligned access is enabled or not is a question of the driver (it really should just always be enabled, but I'm not sure it is). The place to fix that is somewhere else, and not in these low level usage checks

Reply:
Do you mean (1) that `Subtarget->enableFlatScratch()` should be dropped from the OR, or (2) that it should become `(Subtarget->enableFlatScratch() && Subtarget->hasUnalignedAccessMode())` (i.e., it would then depend on whether the mode bit is set, but would not require the UnalignedScratchAccess feature)?

Reply:
Dropped from the or. The unaligned access mode is orthogonal to which scratch ABI is in use.

IIRC the difficulty is linux always enables unaligned access, and windows does not. As an incorrect way of dealing with this, the subtarget constructor assumes unaligned access is enabled for amdhsa (which will be wrong if executed on the compatibility layer on top of pal on windows), and not for amdpal. Don't remember what we do for unknown / mesa3d

Reply:
Done in 2e43c4cc21fc92a80392d899f373f201cbf6ddbc.